src/services/AzureSubscriptionProvider.ts

```diff
@@ -0,0 +1,270 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.md in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+import * as arm from '@azure/arm-subscriptions';
+import { Environment } from '@azure/ms-rest-azure-env';
+import { uiUtils } from '@microsoft/vscode-azext-azureutils';
+import { IActionContext, callWithTelemetryAndErrorHandling } from '@microsoft/vscode-azext-utils';
+import * as vscode from 'vscode';
+import { AzureSubscription } from '../../api/src/index';
+import { settingUtils } from '../utils/settingUtils';
+
+type AzureLoginStatus = 'Initializing' | 'LoggingIn' | 'LoggedIn' | 'LoggedOut';
+export type AzureSubscriptionsResult = {
+    readonly status: AzureLoginStatus;
+    readonly allSubscriptions: AzureSubscription[];
+    readonly filters: AzureSubscription[];
+}
+
+export interface AzureSubscriptionProvider {
+    getSubscriptions(): Promise<AzureSubscriptionsResult>;
+
+    logIn(): Promise<void>;
+    logOut(): Promise<void>;
+    selectSubscriptions(subscriptionIds: string[] | undefined): Promise<void>;
+
+    readonly onStatusChanged: vscode.EventEmitter<AzureLoginStatus>;
+    readonly onFiltersChanged: vscode.EventEmitter<void>;
+    readonly onSessionsChanged: vscode.EventEmitter<void>;
+    readonly onSubscriptionsChanged: vscode.EventEmitter<void>;
+
+    onStatusChangedEvent: vscode.Event<AzureLoginStatus>;
+    onFiltersChangedEvent: vscode.Event<void>;
+    onSessionsChangedEvent: vscode.Event<void>;
+    onSubscriptionsChangedEvent: vscode.Event<void>;
+
+    readonly waitForFilters: () => Promise<boolean>;
+    readonly waitForLogin: () => Promise<boolean>;
+    readonly waitForSubscriptions: () => Promise<boolean>;
+}
+
+export class VSCodeAzureSubscriptionProvider extends vscode.Disposable implements AzureSubscriptionProvider {
```

Comment:
Doesn't have to be in this PR, but IMO we should put this into a package sooner than later. We already _know_ we are going to have to, in order to support extensions like Docker.